File: 백준/Gold/14502. 연구소/연구소.py
import sys, itertools, copy
input = sys.stdin.readline
from collections import deque

n, m = map(int, input().split())
graph = [list(map(int, input().split())) for _ in range(n)]
blank_index = []
dx = [1, 0, -1, 0]
dy = [0, 1, 0, -1]
answer = 0

def make_walls():
    three_walls = list(itertools.combinations(blank_index,3))

    for wall in three_walls:
        for i in range(3):
            x, y = wall[i]
            graph[x][y] = 1

        spread_virus()

        for i in range(3):
            x, y = wall[i]
            graph[x][y] = 0

def spread_virus():
    global answer
    virus_graph = copy.deepcopy(graph)
    queue = deque([])
    for i in range(n):
        for j in range(m):
            if virus_graph[i][j] == 2:
                queue.append((i,j))
    
    while queue:
        x, y = queue.popleft()
        for i in range(4):
            nx, ny = x + dx[i], y + dy[i]
            if 0<=nx<n and 0<=ny<m and virus_graph[nx][ny] == 0:
                virus_graph[nx][ny] = 2
                queue.append((nx,ny))
    
    blank_cnt = 0
    for i in range(n):
        for j in range(m):
            if virus_graph[i][j] == 0:
                blank_cnt += 1
    answer = max(blank_cnt, answer)


for i in range(n):
    for j in range(m):
        if graph[i][j] == 0:
            blank_index.append((i,j))
make_walls()
print(answer)

File: 백준/Gold/14502. 연구소/연구소.js
const input = require('fs').readFileSync(process.platform === 'linux' ? '/dev/stdin' : __dirname + '/input.txt').toString().split('\n')

const [n, m] = input[0].split(' ').map(Number)
const board = []
for (let i = 1; i < n + 1; i++) {
  board.push(input[i].split(' ').map(Number))
}

const candidates = []
const virus = []
for (let i = 0; i < n; i++) {
  for (let j = 0; j < m; j++) {
    if (!board[i][j]) {
      candidates.push([i, j])
    } 
    
    else if (board[i][j] === 2) {
      virus.push([i, j])
    }
  }
}

let answer = 0
const directions = [[0, 1], [0, -1], [1, 0], [-1, 0]]

make_walls(0, 0, [])

console.log(answer)

function make_walls(walls, index) {
  if (walls === 3) {
    const cnt = spread_virus(board.map((row) => [...row]))
    answer = Math.max(answer, cnt)
    return
  }

  for (let i = index; i < candidates.length; i++) {
    const [x, y] = candidates[i]

    board[x][y] = 1
    make_walls(walls + 1, i + 1)
    board[x][y] = 0
  }
}

function spread_virus(board) {
  let idx = 0
  const queue = []

  for (const [x, y] of virus) {
    queue.push([x, y])
  }

  while(idx < queue.length) {
    const [x, y] = queue[idx++]

    for (const [dx, dy] of directions) {
      const [nx, ny] = [dx + x, dy + y]

      if (nx < 0 || ny < 0 || nx >= n || ny >= m) continue
      if (board[nx][ny]) continue

      board[nx][ny] = 2
      queue.push([nx, ny])
    }
  }

  return count_safe_area(board)
}

function count_safe_area(board) {
  let cnt = 0
  for (let i = 0; i < n; i++) {
    for (let j = 0; j < m; j++) {
      if (!board[i][j]) cnt++
    }
  }

  return cnt
}